google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb

```diff
@@ -129,6 +154,60 @@ def stopped?
 
           private
 
+          def handle_failure error, ack_ids, ack_deadline_seconds = nil
+            ack_ids = (parse_error error) || ack_ids
+            perform_retry_async ack_ids, ack_deadline_seconds
+          end
+
+          def parse_error error
+            metadata = error.error_metadata
+            return if metadata.nil?
+            permanent_failures, temporary_failures = metadata.partition do |_, v|
+              v.include? PERMANENT_FAILURE
+            end.map(&:to_h)
+            handle_permanent_failures permanent_failures
+            return temporary_failures.keys.map(&:to_s) unless temporary_failures.empty?
+          end
+
+          def handle_permanent_failures error
+            # TODO: Add log or pass on result(AcknowledgeResult) to callback
+          end
+
+          def perform_retry_async ack_ids, ack_deadline_seconds = nil
+            return unless retry_thread_pool.running?
+            Concurrent::Promises.future_on(
```

Comment:
What happens if the thread pool is shut down between the above two lines?

Reply:
It throws `#<Concurrent::RejectedExecutionError: Concurrent::RejectedExecutionError>` which gets propagated via the error callback registered on subscriber. But this should be a rare scenario as retry_thread_pool would be shutdown only on a stop call on the subscriber and all stream shutdown are waited before shutting the buffer.